File: app/src/main/java/com/robifr/ledger/ui/createcustomer/CreateCustomerFragment.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.createcustomer;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentFactory;
import androidx.lifecycle.ViewModelProvider;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.robifr.ledger.R;
import com.robifr.ledger.databinding.CreateCustomerFragmentBinding;
import com.robifr.ledger.ui.BackStack;
import com.robifr.ledger.ui.FragmentResultKey;
import com.robifr.ledger.ui.createcustomer.viewmodel.CreateCustomerViewModel;
import com.robifr.ledger.util.Compats;
import java.math.BigDecimal;
import java.util.Objects;

public class CreateCustomerFragment extends Fragment implements Toolbar.OnMenuItemClickListener {
  public enum Request implements FragmentResultKey {
    CREATE_CUSTOMER;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  public enum Result implements FragmentResultKey {
    CREATED_CUSTOMER_ID;

    @Override
    @NonNull
    public String key() {
      return FragmentResultKey.generateKey(this);
    }
  }

  @NonNull protected final OnBackPressedHandler _onBackPressed = new OnBackPressedHandler();
  @Nullable protected CreateCustomerFragmentBinding _fragmentBinding;
  @Nullable protected CreateCustomerName _inputName;
  @Nullable protected CreateCustomerBalance _inputBalance;
  @Nullable protected CreateCustomerDebt _inputDebt;

  @Nullable protected CreateCustomerViewModel _createCustomerViewModel;
  @Nullable protected CreateCustomerViewModelHandler _viewModelHandler;

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstance) {
    Objects.requireNonNull(inflater);

    this._fragmentBinding = CreateCustomerFragmentBinding.inflate(inflater, container, false);
    return this._fragmentBinding.getRoot();
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstance) {
    Objects.requireNonNull(view);
    Objects.requireNonNull(this._fragmentBinding);

    this._inputName = new CreateCustomerName(this);
    this._inputBalance = new CreateCustomerBalance(this);
    this._inputDebt = new CreateCustomerDebt(this);
    this._createCustomerViewModel =
        new ViewModelProvider(this, new CreateCustomerViewModel.Factory(this.requireContext()))
            .get(CreateCustomerViewModel.class);
    this._viewModelHandler =
        new CreateCustomerViewModelHandler(this, this._createCustomerViewModel);

    this.requireActivity()
        .getOnBackPressedDispatcher()
        .addCallback(this.getViewLifecycleOwner(), this._onBackPressed);
    this._fragmentBinding.toolbar.getMenu().clear();
    this._fragmentBinding.toolbar.inflateMenu(R.menu.reusable_toolbar_edit);
    this._fragmentBinding.toolbar.setOnMenuItemClickListener(this);
    this._fragmentBinding.toolbar.setNavigationOnClickListener(
        v -> this._onBackPressed.handleOnBackPressed());

    this._createCustomerViewModel.onBalanceChanged(0L);
    this._createCustomerViewModel.onDebtChanged(BigDecimal.ZERO);
  }

  @Override
  public boolean onMenuItemClick(@NonNull MenuItem item) {
    Objects.requireNonNull(item);
    Objects.requireNonNull(this._createCustomerViewModel);

    return switch (item.getItemId()) {
      case R.id.save -> {
        this._createCustomerViewModel.onSave();
        yield true;
      }

      default -> false;
    };
  }

  @NonNull
  public CreateCustomerFragmentBinding fragmentBinding() {
    return Objects.requireNonNull(this._fragmentBinding);
  }

  @NonNull
  public CreateCustomerName inputName() {
    return Objects.requireNonNull(this._inputName);
  }

  @NonNull
  public CreateCustomerBalance inputBalance() {
    return Objects.requireNonNull(this._inputBalance);
  }

  @NonNull
  public CreateCustomerDebt inputDebt() {
    return Objects.requireNonNull(this._inputDebt);
  }

  @NonNull
  public CreateCustomerViewModel createCustomerViewModel() {
    return Objects.requireNonNull(this._createCustomerViewModel);
  }

  public void finish() {
    if (this.requireActivity() instanceof BackStack navigation
        && navigation.currentTabStackTag() != null) {
      Compats.hideKeyboard(this.requireContext(), this.requireView().findFocus());
      navigation.popFragmentStack(navigation.currentTabStackTag());
    }
  }

  public static class Factory extends FragmentFactory {
    @Override
    @NonNull
    public Fragment instantiate(@NonNull ClassLoader classLoader, @NonNull String className) {
      Objects.requireNonNull(classLoader);
      Objects.requireNonNull(className);

      return (className.equals(CreateCustomerFragment.class.getName()))
          ? new CreateCustomerFragment()
          : super.instantiate(classLoader, className);
    }
  }

  private class OnBackPressedHandler extends OnBackPressedCallback {
    public OnBackPressedHandler() {
      super(true);
    }

    @Override
    public void handleOnBackPressed() {
      new MaterialAlertDialogBuilder(CreateCustomerFragment.this.requireContext())
          .setTitle(CreateCustomerFragment.this.getString(R.string.text_discard_this_unsaved_task))
          .setNegativeButton(
              CreateCustomerFragment.this.getString(R.string.text_discard),
              (dialog, type) -> CreateCustomerFragment.this.finish())
          .setPositiveButton(
              CreateCustomerFragment.this.getString(R.string.text_cancel),
              (dialog, type) -> dialog.dismiss())
          .show();
    }
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createcustomer/CreateCustomerFragment.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createcustomer

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updatePadding
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import com.robifr.ledger.R
import com.robifr.ledger.databinding.CreateCustomerFragmentBinding
import com.robifr.ledger.ui.FragmentResultKey
import com.robifr.ledger.ui.OnBackPressedHandler
import com.robifr.ledger.ui.SnackbarState
import com.robifr.ledger.ui.createcustomer.viewmodel.CreateCustomerEvent
import com.robifr.ledger.ui.createcustomer.viewmodel.CreateCustomerResultState
import com.robifr.ledger.ui.createcustomer.viewmodel.CreateCustomerState
import com.robifr.ledger.ui.createcustomer.viewmodel.CreateCustomerViewModel
import com.robifr.ledger.ui.createcustomer.viewmodel.CustomerBalanceAddState
import com.robifr.ledger.ui.createcustomer.viewmodel.CustomerBalanceWithdrawState
import com.robifr.ledger.util.hideKeyboard
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
open class CreateCustomerFragment : Fragment(), Toolbar.OnMenuItemClickListener {
  protected var _fragmentBinding: CreateCustomerFragmentBinding? = null
  val fragmentBinding: CreateCustomerFragmentBinding
    get() = _fragmentBinding!!

  open val createCustomerViewModel: CreateCustomerViewModel by viewModels()
  private lateinit var _inputName: CreateCustomerName
  private lateinit var _inputBalance: CreateCustomerBalance
  private lateinit var _inputDebt: CreateCustomerDebt

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View {
    _fragmentBinding = CreateCustomerFragmentBinding.inflate(inflater, container, false)
    _inputName = CreateCustomerName(this)
    _inputBalance = CreateCustomerBalance(this)
    _inputDebt = CreateCustomerDebt(this)
    return fragmentBinding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    ViewCompat.setOnApplyWindowInsetsListener(fragmentBinding.root) { view, insets ->
      val systemBarInsets: Insets = insets.getInsets(WindowInsetsCompat.Type.systemBars())
      val windowInsets: Insets =
          insets.getInsets(
              WindowInsetsCompat.Type.displayCutout() or WindowInsetsCompat.Type.navigationBars())
      view.updatePadding(
          top = systemBarInsets.top, left = windowInsets.left, right = windowInsets.right)
      WindowInsetsCompat.CONSUMED
    }
    requireActivity()
        .onBackPressedDispatcher
        .addCallback(
            viewLifecycleOwner, OnBackPressedHandler { createCustomerViewModel.onBackPressed() })
    fragmentBinding.toolbar.setNavigationOnClickListener { createCustomerViewModel.onBackPressed() }
    fragmentBinding.toolbar.menu.clear()
    fragmentBinding.toolbar.inflateMenu(R.menu.reusable_toolbar_edit)
    fragmentBinding.toolbar.setOnMenuItemClickListener(this)
    createCustomerViewModel.uiEvent.observe(viewLifecycleOwner, ::_onUiEvent)
    createCustomerViewModel.uiState.observe(viewLifecycleOwner, ::_onUiState)
    createCustomerViewModel.balanceView.addBalanceState.observe(
        viewLifecycleOwner, ::_onInputBalanceAmountState)
    createCustomerViewModel.balanceView.withdrawBalanceState.observe(
        viewLifecycleOwner, ::_onInputWithdrawAmountState)
  }

  override fun onMenuItemClick(item: MenuItem?): Boolean =
      when (item?.itemId) {
        R.id.save -> {
          createCustomerViewModel.onSave()
          true
        }
        else -> false
      }

  fun finish() {
    requireView().hideKeyboard()
    findNavController().popBackStack()
  }

  private fun _onUiEvent(event: CreateCustomerEvent) {
    event.snackbar?.let {
      _onSnackbarState(it.data)
      it.onConsumed()
    }
    event.isFragmentFinished?.let {
      finish()
      it.onConsumed()
    }
    event.isUnsavedChangesDialogShown?.let {
      _onUnsavedChangesDialogState(onDismiss = { it.onConsumed() })
    }
    event.createResult?.let {
      _onResultState(it.data)
      it.onConsumed()
    }
  }

  private fun _onResultState(state: CreateCustomerResultState) {
    parentFragmentManager.setFragmentResult(
        Request.CREATE_CUSTOMER.key(),
        Bundle().apply {
          state.createdCustomerId?.let { putLong(Result.CREATED_CUSTOMER_ID_LONG.key(), it) }
        })
    finish()
  }

  private fun _onSnackbarState(state: SnackbarState) {
    Snackbar.make(
            fragmentBinding.root as View,
            state.messageRes.toStringValue(requireContext()),
            Snackbar.LENGTH_LONG)
        .show()
  }

  private fun _onUnsavedChangesDialogState(onDismiss: () -> Unit) {
    MaterialAlertDialogBuilder(requireContext())
        .setMessage(R.string.createCustomer_unsavedChangesWarning)
        .setNegativeButton(R.string.action_cancel) { _, _ -> }
        .setPositiveButton(R.string.action_leaveWithoutSaving) { _, _ -> finish() }
        .setOnDismissListener { onDismiss() }
        .show()
  }

  private fun _onUiState(state: CreateCustomerState) {
    _inputName.setInputtedNameText(
        state.name, state.nameErrorMessageRes?.toStringValue(requireContext()))
    _inputBalance.setInputtedBalance(state.balance)
    _inputBalance.setAddBalanceButtonEnabled(state.isAddBalanceButtonEnabled)
    _inputBalance.setWithdrawBalanceButtonEnabled(state.isWithdrawBalanceButtonEnabled)
    _inputDebt.setInputtedDebt(state.debt, state.debtColorRes)
  }

  private fun _onInputBalanceAmountState(state: CustomerBalanceAddState) {
    if (state.isDialogShown) _inputBalance.showAddBalanceDialog()
    else _inputBalance.dismissAddBalanceDialog()
    _inputBalance.setInputtedBalanceAmountText(state.formattedAmount, state.isAddButtonEnabled)
  }

  private fun _onInputWithdrawAmountState(state: CustomerBalanceWithdrawState) {
    if (state.isDialogShown) _inputBalance.showWithdrawBalanceDialog()
    else _inputBalance.dismissWithdrawBalanceDialog()
    _inputBalance.setInputtedWithdrawAmountText(
        state.formattedAmount, state.isWithdrawButtonEnabled)
    _inputBalance.setAvailableAmountToWithdraw(state.availableAmountToWithdraw)
  }

  enum class Request : FragmentResultKey {
    CREATE_CUSTOMER
  }

  enum class Result : FragmentResultKey {
    CREATED_CUSTOMER_ID_LONG
  }
}
